astropy/stats/funcs.py

```diff
@@ -681,15 +681,29 @@ def median_absolute_deviation(a, axis=None):
 
     """
 
-    a = np.array(a, copy=False)
-    a_median = np.median(a, axis=axis)
+    # Check if the array has a mask and if so use np.ma.median
+    # See https://github.com/numpy/numpy/issues/7330 why using np.ma.median
+    # for normal arrays should not be done (summary: np.ma.median always
+    # returns an masked array even if the result should be scalar).
+    try:
+        if np.any(a.mask):
+            func = np.ma.median
+        else:
+            # Array has a mask but no value is masked so we can savely use
+            # np.median on it.
+            func = np.median
+    except AttributeError:
+        func = np.median
+        a = np.array(a, copy=False)
```

Comment:
I realise you are just following the original, but I'd use `np.asanyarray` here since that would allow one to pass in quantities as well (`np.median` works on those).